src/Adapter/ContainerBuilder.php

```diff
@@ -38,35 +45,177 @@
+
+        //These methods load required files like autoload or annotation metadata so we need to load
+        //them at each container creation, this can't be compiled.
+        $this->loadDoctrineAnnotationMetadata();
+        $this->loadModulesAutoloader();
+
+        $container = $this->loadDumpedContainer();
+        if (null === $container) {
+            $container = $this->compileContainer();
+        }
+
+        return $container;
+    }
+
+    /**
+     * @return ContainerInterface|null
+     */
+    private function loadDumpedContainer()
+    {
+        $container = null;
+        if ($this->containerConfigCache->isFresh()) {
+            require_once $this->dumpFile;
+            $container = new $this->containerClassName();
         }
 
+        return $container;
+    }
+
+    /**
+     * @return SfContainerBuilder
+     *
+     * @throws \Exception
+     */
+    private function compileContainer()
+    {
         $container = new SfContainerBuilder();
+
         $container->addCompilerPass(new LegacyCompilerPass());
-        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
-        $env = $isDebug ? 'dev' : 'prod';
-        $servicesPath = _PS_CONFIG_DIR_ . "services/${name}/services_${env}.yml";
-        $loader->load($servicesPath);
+        $this->loadServices($container);
+
+        //Build extensions
+        $builderExtensions = [
+            new DoctrineBuilderExtension($this->environment),
+        ];
+        /** @var ContainerBuilderExtensionInterface $builderExtension */
+        foreach ($builderExtensions as $builderExtension) {
+            $builderExtension->build($container);
+        }
+
         $container->compile();
 
+        //Dump the container file
         $dumper = new PhpDumper($container);
-        file_put_contents($file, $dumper->dump(array('class' => $containerName)));
+        $this->containerConfigCache->write(
+            $dumper->dump(['class' => $this->containerClassName]),
+            $container->getResources()
+        );
 
         return $container;
     }
+
+    /**
+     * In symfony context doctrine classes (like Table, Entity, ...) are available thanks to
+     * the autoloader. In this specific context we don't have the general autoloader, so we need
+     * to include these classes manually. This is performed in Doctrine\ORM\Configuration::newDefaultAnnotationDriver
+     * which is called in Setup::createAnnotationMetadataConfiguration.
+     */
+    private function loadDoctrineAnnotationMetadata()
+    {
+        Setup::createAnnotationMetadataConfiguration([]);
+    }
+
+    /**
+     * @param SfContainerBuilder $container
+     *
+     * @throws \Exception
+     */
+    private function loadServices(SfContainerBuilder $container)
+    {
+        $loader = new YamlFileLoader($container, new FileLocator(__DIR__));
+        $servicesPath = _PS_CONFIG_DIR_ . sprintf('services/%s/services_%s.yml', $this->containerName, $this->environment->getName());
```

Comment:
```suggestion
        $servicesPath = sprintf(
            '%sservices/%s/services_%s.yml',
            _PS_CONFIG_DIR_,
            $this->containerName, 
            $this->environment->getName()
        );
```

Reply:
I would agree with you in general but I don't like this notation here because the first `%s` is kinda hidden. If there was a `/` right after I wouldn't mind But in this case I think it can be confusing when someone reads it for the first time

Reply:
it's just the C syntax :)
I think it's less confusing than mixin concat & sprintf :/

Reply:
Alright let's go for it